stetho/src/main/java/com/facebook/stetho/inspector/protocol/module/DOMStorage.java

```diff
@@ -2,20 +2,185 @@
+import com.facebook.stetho.inspector.domstorage.SharedPreferencesHelper;
 import com.facebook.stetho.inspector.jsonrpc.JsonRpcPeer;
+import com.facebook.stetho.inspector.jsonrpc.JsonRpcResult;
 import com.facebook.stetho.inspector.protocol.ChromeDevtoolsDomain;
 import com.facebook.stetho.inspector.protocol.ChromeDevtoolsMethod;
 
+import com.facebook.stetho.json.ObjectMapper;
+import com.facebook.stetho.json.annotation.JsonProperty;
+import org.json.JSONException;
 import org.json.JSONObject;
 
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
 public class DOMStorage implements ChromeDevtoolsDomain {
-  public DOMStorage() {
+  private final Context mContext;
+  private final DOMStoragePeerManager mDOMStoragePeerManager;
+  private final ObjectMapper mObjectMapper = new ObjectMapper();
+
+  public DOMStorage(Context context) {
+    mContext = context;
+    mDOMStoragePeerManager = new DOMStoragePeerManager(context);
   }
 
   @ChromeDevtoolsMethod
   public void enable(JsonRpcPeer peer, JSONObject params) {
+    mDOMStoragePeerManager.addPeer(peer);
   }
 
+  @ChromeDevtoolsMethod
   public void disable(JsonRpcPeer peer, JSONObject params) {
+    mDOMStoragePeerManager.removePeer(peer);
+  }
+
+  @ChromeDevtoolsMethod
+  public JsonRpcResult getDOMStorageItems(JsonRpcPeer peer, JSONObject params)
+      throws JSONException {
+    StorageId storage = mObjectMapper.convertValue(
+        params.getJSONObject("storageId"),
+        StorageId.class);
+
+    ArrayList<List<String>> entries = new ArrayList<List<String>>();
+    String prefTag = storage.securityOrigin;
+    if (storage.isLocalStorage) {
+      SharedPreferences prefs = mContext.getSharedPreferences(prefTag, Context.MODE_PRIVATE);
+      for (Map.Entry<String, ?> prefsEntry : prefs.getAll().entrySet()) {
+        ArrayList<String> entry = new ArrayList<String>(2);
+        entry.add(prefsEntry.getKey());
+        entry.add(SharedPreferencesHelper.valueToString(prefsEntry.getValue()));
+        entries.add(entry);
+      }
+    }
+
+    GetDOMStorageItemsResult result = new GetDOMStorageItemsResult();
+    result.entries = entries;
+    return result;
+  }
+
+  @ChromeDevtoolsMethod
+  public void setDOMStorageItem(JsonRpcPeer peer, JSONObject params) throws JSONException {
+    StorageId storage = mObjectMapper.convertValue(
+        params.getJSONObject("storageId"),
+        StorageId.class);
+    String key = params.getString("key");
+    String value = params.getString("value");
+
+    if (storage.isLocalStorage) {
+      SharedPreferences prefs = mContext.getSharedPreferences(
+          storage.securityOrigin,
+          Context.MODE_PRIVATE);
+      Object exitingValue = prefs.getAll().get(key);
+      SharedPreferences.Editor editor = prefs.edit();
+      if (!tryAssignByType(editor, key, value, exitingValue)) {
+        editor.putString(key, value);
```

Comment:
This seems pretty fragile, and a good way to break the app. 

I dont have a great way to test this, but when does `setDOMStorageItem` get called on the devtools side?  If it is when you leave the field this has the ability to be pretty harmful to the hosting app.There is no validation, and no telling what the app expects this to be.